internal/featuretests/authorization_test.go

```diff
@@ -0,0 +1,568 @@
+		TypeUrl:   listenerType,
+		Resources: resources(t, staticListener()),
+	}).Status(invalid).Equals(projcontour.HTTPProxyStatus{
+		CurrentStatus: k8s.StatusInvalid,
+		Description:   `Spec.Virtualhost.Authorization.ExtensionServiceRef specifies an unsupported resource version "foo/bar"`,
+	})
+
+	invalid.Spec.VirtualHost.Authorization.ServiceRef = projcontour.ExtensionServiceReference{
+		APIVersion: "projectcontour.io/v1alpha1",
+		Kind:       "Foo",
+		Namespace:  "",
+		Name:       "",
+	}
+
+	rh.OnDelete(invalid)
+	rh.OnAdd(invalid)
+
+	c.Request(listenerType).Equals(&v2.DiscoveryResponse{
+		TypeUrl:   listenerType,
+		Resources: resources(t, staticListener()),
+	}).Status(invalid).Equals(projcontour.HTTPProxyStatus{
+		CurrentStatus: k8s.StatusInvalid,
+		Description:   `Spec.Virtualhost.Authorization.ExtensionServiceRef specifies an unsupported resource kind "Foo"`,
+	})
+
+	invalid.Spec.VirtualHost.Authorization.ServiceRef = projcontour.ExtensionServiceReference{
+		APIVersion: "projectcontour.io/v1alpha1",
+		Kind:       "ExtensionService",
+		Namespace:  "missing",
+		Name:       "extension",
+	}
+
+	rh.OnDelete(invalid)
+	rh.OnAdd(invalid)
+
+	c.Request(listenerType).Equals(&v2.DiscoveryResponse{
+		TypeUrl:   listenerType,
+		Resources: resources(t, staticListener()),
+	}).Status(invalid).Equals(projcontour.HTTPProxyStatus{
+		CurrentStatus: k8s.StatusInvalid,
+		Description:   `Spec.Virtualhost.Authorization.ExtensionServiceRef support service "missing/extension" not found`,
+	})
+
+	invalid.Spec.VirtualHost.Authorization.ServiceRef = projcontour.ExtensionServiceReference{
+		Namespace: "auth",
+		Name:      "extension",
+	}
+
+	rh.OnDelete(invalid)
+	rh.OnAdd(invalid)
+
+	c.Request(listenerType).Equals(&v2.DiscoveryResponse{
+		TypeUrl: listenerType,
+		Resources: resources(t,
+			defaultHTTPListener(),
+			&v2.Listener{
+				Name:    "ingress_https",
+				Address: envoy.SocketAddress("0.0.0.0", 8443),
+				ListenerFilters: envoy.ListenerFilters(
+					envoy.TLSInspector(),
+				),
+				FilterChains: []*envoy_api_v2_listener.FilterChain{
+					filterchaintls(fqdn,
+						&corev1.Secret{
+							ObjectMeta: fixture.ObjectMeta("certificate"),
+							Type:       "kubernetes.io/tls",
+							Data:       secretdata(CERTIFICATE, RSA_PRIVATE_KEY),
+						},
+						authzFilterFor(
+							fqdn,
+							&envoy_config_filter_http_ext_authz_v2.ExtAuthz{
+								Services:               grpcCluster("extension/auth/extension"),
+								ClearRouteCache:        true,
+								FailureModeAllow:       false,
+								IncludePeerCertificate: true,
+								StatusOnError: &envoy_type.HttpStatus{
+									Code: envoy_type.StatusCode_Forbidden,
+								},
+							},
+						),
+						nil, "h2", "http/1.1"),
+				},
+				SocketOptions: envoy.TCPKeepaliveSocketOptions(),
+			},
+			staticListener()),
+	}).Status(invalid).Like(projcontour.HTTPProxyStatus{
+		CurrentStatus: k8s.StatusValid,
+	})
+}
+
+func TestAuthorization(t *testing.T) {
```

Comment:
No problems here, just really like this approach to making the feature tests more readable.

Reply:
The loop variable aliasing bit me hard though; I wonder if it affects subtests generally?